src/Interpreters/sortBlock.cpp

```diff
@@ -32,67 +23,53 @@ struct ColumnWithSortDescription
     /// It means, that this column is ColumnConst
     bool column_const = false;
 };
+
 using ColumnsWithSortDescriptions = std::vector<ColumnWithSortDescription>;
 
-void flattenTupleColumnRecursively(
-    ColumnsWithSortDescriptions & res, const ColumnTuple * tuple, const SortColumnDescription & description, bool has_collation)
+namespace
 {
-    for (const auto & column : tuple->getColumns())
-    {
-        if (const auto * subtuple = typeid_cast<const ColumnTuple *>(column.get()))
-            flattenTupleColumnRecursively(res, subtuple, description, has_collation);
-        else
-        {
-            res.emplace_back(ColumnWithSortDescription{column.get(), description, isColumnConst(*column)});
-            if (has_collation && !res.back().column->isCollationSupported())
-                res.back().description.collator = nullptr;
-        }
-    }
-}
 
-ColumnsWithSortDescriptions getColumnsWithSortDescription(const Block & block, const SortDescription & description)
+inline bool isCollationRequired(const SortColumnDescription & description)
 {
-    size_t size = description.size();
-    ColumnsWithSortDescriptions res;
-    res.reserve(size);
-
-    for (size_t i = 0; i < size; ++i)
-    {
-        const IColumn * column = !description[i].column_name.empty()
-            ? block.getByName(description[i].column_name).column.get()
-            : block.safeGetByPosition(description[i].column_number).column.get();
-
-        if (isCollationRequired(description[i]))
-        {
-            if (!column->isCollationSupported())
-                throw Exception(
-                    "Collations could be specified only for String, LowCardinality(String), Nullable(String) or for Array or Tuple, "
-                    "containing them.",
-                    ErrorCodes::BAD_COLLATION);
-        }
-        if (const auto * tuple = typeid_cast<const ColumnTuple *>(column))
-            flattenTupleColumnRecursively(res, tuple, description[i], isCollationRequired(description[i]));
-        else
-            res.emplace_back(ColumnWithSortDescription{column, description[i], isColumnConst(*column)});
-    }
-    return res;
+    return description.collator != nullptr;
 }
 
-struct PartialSortingLess
+template <bool check_collation>
+struct PartialSortingLessImpl
 {
     const ColumnsWithSortDescriptions & columns;
 
-    explicit PartialSortingLess(const ColumnsWithSortDescriptions & columns_) : columns(columns_) {}
+    explicit PartialSortingLessImpl(const ColumnsWithSortDescriptions & columns_) : columns(columns_) { }
 
-    bool operator() (size_t a, size_t b) const
+    inline bool operator()(size_t a, size_t b) const
```

Comment:
Why do we need `inline` here? The code is already inlined by definition.

Reply:
@amosbird not sure, why it is inlined by definition ?

Reply:
@amosbird seems like this rule is defined by the standard, but to increase some inline function `score` most of the time inline is written explicitly. It is not necessary, because to force inline only attribute should be used.

Reply:
> but to increase some inline function score most of the time inline is written explicitly

Could you provide some reference (gcc or clang) about this `score` thing? I'm curious how it works.

Reply:
@amosbird I am not sure about this, it is just assumption. Compiler can choose inline or not to inline member function, if inline is explicitly specified I would assume that there are more changes that function will get inlined by compiler.
Anyway if you mark function inline it can be as some sort of documentation for reader, for example in our code base I see that boost in some places mark member functions inline explicitly.